extensions.html

```diff
@@ -182,19 +184,142 @@ <h2>
+        A {{GamepadHapticActuator}} corresponds to a configuration of motors or
+        other actuators that can apply a force for the purposes of haptic
+        feedback.
       </p>
       <pre class="idl">
         [Exposed=Window]
         interface GamepadHapticActuator {
           readonly attribute GamepadHapticActuatorType type;
+          boolean canPlayEffectType(GamepadHapticEffectType type);
+          Promise&lt;GamepadHapticsResult&gt; playEffect(
+              GamepadHapticEffectType type,
+              optional GamepadEffectParameters params = {});
           Promise&lt;boolean&gt; pulse(double value, double duration);
+          Promise&lt;GamepadHapticsResult&gt; reset();
         };
       </pre>
+      <p>
+        Instances of {{GamepadHapticActuator}} are created with the internal
+        slots described in the following table:
+      </p>
+      <table class="simple">
+        <tr>
+          <th>
+            Internal slot
+          </th>
+          <th>
+            Initial value
+          </th>
+          <th>
+            Description (non-normative)
+          </th>
+        </tr>
+        <tr>
+          <td>
+            <dfn data-dfn-for=
+            "GamepadHapticActuator">[[\playingEffectPromise]]</dfn>
+          </td>
+          <td>
+            `undefined`
+          </td>
+          <td>
+            The {{Promise}} for the playing effect, or `undefined` if no effect
+            is playing.
+          </td>
+        </tr>
+      </table>
+      <dl>
+        <dt>
+          <dfn>type</dfn> attribute
+        </dt>
+        <dd>
+          <p>
+            A {{GamepadHapticActuatorType}} value representing the actuator
+            type.
+          </p>
+        </dd>
+      </dl>
       <dl data-dfn-for="GamepadHapticActuator">
         <dt>
-          <dfn>pulse()</dfn>
+          <dfn>canPlayEffectType()</dfn> method
+        </dt>
+        <dd>
+          <p>
+            The {{GamepadHapticActuator/canPlayEffectType()}} method steps are:
+          </p>
+          <ol>
+            <li>If this actuator can [=play effects with type=]
+            |type:GamepadHapticEffectType|, return `true`.
+            </li>
+            <li>Otherwise, return `false`.
+            </li>
+          </ol>
+        </dd>
+        <dt>
+          <dfn>playEffect()</dfn> method
+        </dt>
+        <dd>
+          <p>
+            The {{GamepadHapticActuator/playEffect()}} method steps are:
+          </p>
+          <ol>
+            <li>If |params:GamepadEffectParameters| does not describe a [=valid
+            effect=] of type |type:GamepadHapticEffectType|, return [=a promise
+            resolved with=] {{GamepadHapticsResult/"invalid-parameter"}}.
+            </li>
+            <li>Let |document| be the [=current settings object=]'s [=relevant
+            global object=]'s [=associated `Document`=].
+            </li>
+            <li>If |document| is `null` or |document| is not [=Document/fully
```

Comment:
Same here...

Reply:
Which error would you use?

I feel a preempted effect is not an error because it's expected you'll sometimes interrupt an effect to cut it short or replace it with a new effect. I think visibility changes should be handled as if the effect were preempted by a call to `reset()`.

Reply:
Understood. The case here feels different because this step is happening as a result of an explicit call to `.playEffect()`. The change in document visibility resolves the `[[playingEffectPromise]]` with "preempted", but that's different from this call. 

Note: in addition the visibility change handler below, you can add some steps to watch for the document becoming not fully active. The visibility change and the document not being fully active can then have the same behavior.

Reply:
Also, with `AbortController` support, then `.reset()` can effectively go away. It's all controller from a single place.

Reply:
If I understand correctly this would be a new (optional) effect param and would let you do something like this:

    const controller = new AbortController();
    const playEffectPromise = gamepad.vibrationActuator.playEffect('dual-rumble', {signal:controller.signal, strongMagnitude:1.0, duration:1000});
    controller.abort(); // instead of gamepad.vibrationActuator.reset()
    const effectResult = await playEffectPromise; // resolves to 'preempted'

That sounds good to me.